src/packet/bwe/loss_controller.rs

```diff
@@ -0,0 +1,1341 @@
+            );
+            dbg!(self.state, state);
+        }
+        self.state = state;
+    }
+}
+
+#[derive(Debug)]
+struct PacketResultsSummary {
+    num_packets: u64,
+    num_lost_packets: u64,
+    total_size: u64,
+    first_send_time: Instant,
+    last_send_time: Instant,
+}
+
+impl PacketResultsSummary {
+    pub fn new(first_send_time: Instant, last_send_time: Instant) -> PacketResultsSummary {
+        PacketResultsSummary {
+            num_packets: 0,
+            num_lost_packets: 0,
+            total_size: 0,
+            last_send_time,
+            first_send_time,
+        }
+    }
+
+    pub fn from(records: &[TwccSendRecord]) -> Option<PacketResultsSummary> {
+        let first = records.first()?;
+
+        let mut summary =
+            PacketResultsSummary::new(first.local_send_time(), first.local_send_time());
+        for record in records {
+            summary.num_packets += 1;
+            summary.total_size += record.size() as u64;
+            summary.first_send_time = min(summary.first_send_time, record.local_send_time());
+            summary.last_send_time = max(summary.last_send_time, record.local_send_time());
+
+            if record.remote_recv_time().is_none() {
+                summary.num_lost_packets += 1;
+            }
+        }
+
+        Some(summary)
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct Observation {
+    num_packets: u64,
+    num_lost_packets: u64,
+    num_received_packets: u64,
+    sending_rate: Bitrate,
+    id: u64,
+    is_initialized: bool,
+}
+
+impl Observation {
+    pub const DUMMY: Self = Self {
+        num_packets: 0,
+        num_lost_packets: 0,
+        num_received_packets: 0,
+        sending_rate: Bitrate::NEG_INFINITY,
+        id: 0,
+        is_initialized: false,
+    };
+
+    pub fn with(partial_observation: &PartialObservation, id: u64) -> Observation {
+        Observation {
+            num_packets: partial_observation.num_packets,
+            num_lost_packets: partial_observation.num_lost_packets,
+            num_received_packets: partial_observation.num_packets
+                - partial_observation.num_lost_packets,
+            sending_rate: Bitrate::from(partial_observation.size as f64 / 1.0),
+            id,
+            is_initialized: true,
+        }
+    }
+}
+
+struct PartialObservation {
+    num_packets: u64,
+    num_lost_packets: u64,
+    size: u64,
+}
+
+impl PartialObservation {
+    pub fn new() -> PartialObservation {
+        PartialObservation {
+            num_packets: 0,
+            num_lost_packets: 0,
+            size: 0,
+        }
+    }
+
+    pub fn update(&mut self, summary: PacketResultsSummary) {
+        // TODO: figure out whether to use usize or u64
+        self.num_packets += summary.num_packets;
+        self.num_lost_packets += summary.num_lost_packets;
+        self.size += summary.total_size;
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct ChannelParameters {
+    inherent_loss: f64,
```

Comment:
inherent_loss, is it fraction or percent? either in name or newtype?

Reply:
It's percent expressed as a value in the range `[0, 1.0]`